pkg/v1/tkg/managementcomponents/management_component_install.go

```diff
@@ -48,18 +60,202 @@ type ManagementComponentsInstallOptions struct {
 	ManagementPackageRepositoryOptions ManagementPackageRepositoryOptions
 }
 
+func generateAddonSecretName(clusterName, addonName string) string {
+	return fmt.Sprintf("%s-tanzu-%s-addon", clusterName, addonName)
+}
+
+func pauseAddonSecretReconciliation(clusterClient clusterclient.Client, addonSecreteName, namespace string) error {
+	log.Infof("Pausing reconciliation for %s/%s secret", namespace, addonSecreteName)
+	secret := &corev1.Secret{}
+	jsonPatch := []map[string]interface{}{
+		{
+			"op":    "add",
+			"path":  fmt.Sprintf("/metadata/annotations/tkg.tanzu.vmware.com~1addon-paused"),
+			"value": "",
+		},
+	}
+	payloadBytes, err := json.Marshal(jsonPatch)
+	if err != nil {
+		return errors.Wrap(err, "unable to generate json patch")
+	}
+
+	err = clusterClient.PatchResource(secret, addonSecreteName, namespace, string(payloadBytes), types.JSONPatchType, nil)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return errors.Wrapf(err, "failed to pause %s secret reconciliation", addonSecreteName)
+	}
+	return nil
+}
+
+func pausePackageInstallReconciliation(clusterClient clusterclient.Client, pkgiName, namespace string) error {
+	log.Infof("Pausing reconciliation for %s/%s packageinstall", namespace, pkgiName)
+	pkgi := &kappipkg.PackageInstall{}
+	jsonPatch := []map[string]interface{}{
+		{
+			"op":    "add",
+			"path":  "/spec/paused",
+			"value": true,
+		},
+	}
+	payloadBytes, err := json.Marshal(jsonPatch)
+	if err != nil {
+		return errors.Wrap(err, "unable to generate json patch")
+	}
+	err = clusterClient.PatchResource(pkgi, pkgiName, namespace, string(payloadBytes), types.JSONPatchType, nil)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return errors.Wrapf(err, "failed to pause %s packageinstall reconciliation", pkgiName)
+	}
+	return nil
+}
+
+// PausedAddonLifecycleManagement pauses/unpauses the lifecycle management of addon package with given name and namespace
+func PauseAddonLifecycleManagement(clusterClient clusterclient.Client, clusterName, addonName, namespace string) error {
+	log.Infof("Pausing lifecycle management for %s", addonName)
+	pkgiName := fmt.Sprintf("tanzu-%s", addonsManagerName)
+	addonSecretName := generateAddonSecretName(clusterName, addonsManagerName)
+
+	err := pauseAddonSecretReconciliation(clusterClient, addonSecretName, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = pausePackageInstallReconciliation(clusterClient, pkgiName, namespace)
+	if err != nil {
+		return err
```

Comment:
If we get here and this error occurs, that means the secret reconciliation has been paused, but presumably on the package install? Are there remediation steps? what happens if someone tries the operation again? Will this fail at the pauseAddonSecret step?

Reply:
If we fail at this point, then the remediation step would be to retry the upgrade. 
If total failure happens and user changes their mind about upgrade, they will have to unpause the addons secret reconciliation by hand.  We could try to "roll back" the addons secret pause, but it is difficult to predict the state of the system with enough accuracy to ensure a succesfull "roll back" . Perhaps is best to let the user explore why the failure and unpause  by hand. We do put out a log that we are pausing the addons secret.  Perhaps we should put a log here saying that user needs to check system and unpause secret. 

What will happen then is that the addonsecret is paused so there is no reconcilation of old addons-manager. 
This would not block any new attempts to upgrade. We tested locally that scenario and the upgrade is "succesful" if you restart it after a failure at this point. 

The only other scenario is if the user decides to not do the upgrade after a failure here, which means they would have to manually unpause the lifecycle.

Reply:
Ok, then it probably makes sense to add that log message when this second pause fails, and then filing a docs bug to make sure that the steps for recovery are documented when a user doesn't just retry the upgrade.